File: constrained-app/src/model/constraint.ts
import type { Point } from "./primitive";

export type Constraint = PointsMatch | PointFixed | WidthFixed | HeightFixed;

export class PointsMatch {
    first: number;
    second: number;
    firstPoint: RectPoint;
    secondPoint: RectPoint;

    constructor(first: number, second: number, firstPoint: RectPoint, secondPoint: RectPoint) {
        this.first = first;
        this.second = second;
        this.firstPoint = firstPoint;
        this.secondPoint = secondPoint;
    }

    toJSON() {
        return {
            PointsMatch: {
                first: this.first,
                second: this.second,
                firstPoint: this.firstPoint,
                secondPoint: this.secondPoint
            }
        };
    }
}

export class PointFixed {
    rect: number;
    point: RectPoint;
    target: Point;

    constructor(rect: number, point: RectPoint, target: Point) {
        this.rect = rect;
        this.point = point;
        this.target = target;
    }

    toJSON() {
        return {
            PointFixed: {
                rect: this.rect,
                point: this.point,
                target: this.target
            }
        };
    }
}

export class WidthFixed {
    rect: number;
    width: number;

    constructor(rect: number, width: number) {
        this.rect = rect;
        this.width = width;
    }

    toJSON() {
        return {
            WidthFixed: {
                rect: this.rect,
                width: this.width
            }
        };
    }
}

export class HeightFixed {
    rect: number;
    height: number;

    constructor(rect: number, height: number) {
        this.rect = rect;
        this.height = height;
    }

    toJSON() {
        return {
            HeightFixed: {
                rect: this.rect,
                height: this.height
            }
        };
    }
}

export enum RectPoint {
    TopLeft = "TopLeft",
    TopRight = "TopRight",
    BottomLeft = "BottomLeft",
    BottomRight = "BottomRight"
}

File: constrained-app/src-tauri/src/constraint.rs
use crate::primitive::Point;

#[derive(serde::Serialize, serde::Deserialize)]
pub(crate) enum Constraint {
  PointsMatch {
    first: usize,
    second: usize,
    #[serde(rename = "firstPoint")]
    first_point: RectPoint,
    #[serde(rename = "secondPoint")]
    second_point: RectPoint,
  },
  PointFixed {
    rect: usize,
    point: RectPoint,
    target: Point,
  },
  WidthFixed {
    rect: usize,
    width: f32,
  },
  HeightFixed {
    rect: usize,
    height: f32,
  },
}

impl From<Constraint> for constrained_core::Constraint {
  fn from(constraint: Constraint) -> Self {
    match constraint {
      Constraint::PointsMatch {
        first,
        second,
        first_point,
        second_point,
      } => Self::PointsMatch {
        first,
        second,
        first_point: first_point.into(),
        second_point: second_point.into(),
      },
      Constraint::PointFixed {
        rect,
        point,
        target,
      } => Self::PointFixed {
        rect,
        point: point.into(),
        target: target.into(),
      },
      Constraint::WidthFixed { rect, width } => Self::WidthFixed { rect, width },
      Constraint::HeightFixed { rect, height } => Self::HeightFixed { rect, height },
    }
  }
}

#[derive(serde::Serialize, serde::Deserialize)]
pub enum RectPoint {
  TopLeft,
  TopRight,
  BottomLeft,
  BottomRight,
}

impl From<RectPoint> for constrained_core::RectPoint {
  fn from(point: RectPoint) -> Self {
    match point {
      RectPoint::TopLeft => Self::TopLeft,
      RectPoint::TopRight => Self::TopRight,
      RectPoint::BottomLeft => Self::BottomLeft,
      RectPoint::BottomRight => Self::BottomRight,
    }
  }
}
